### PR TITLE
Check if the php module 'pcntl' is available for occ command processing

### PR DESCRIPTION
**Pull request by mmattel**
This PR notifies via the admin page if the php module `pcntl` is available for occ commands.
Raised at: https://github.com/owncloud/core/pull/21701

image

I will also add a documentation PR.

@DeepDiver1975 @PVince81 @oparoz @rullzer

### CODE REVIEW

**Comment by karlitschek** [comment]
does someone know which linux distributions are affected my this? so which distros ship php without this by default?

**Comment by PVince81** [comment]
@karlitschek happened on my openSUSE Tumbleweed

**Comment by karlitschek** [comment]
can we make this a black hint instead of a red hard error ?

**Comment by mmattel** [comment]
From what I have identified, list to be expanded:
included: Ubuntu 14.04
not included: openSUSE Tumbleweed
others: to be checked, but now easy because of this notification...

@karlitschek @PVince81 
Coloring: sure if you let me know how to change the color or the location. I do not know that...

File: settings/admin.php [hunk 1]
```diff
@@ -137,6 +137,11 @@
 $path = getenv('PATH');
 $template->assign('getenvServerNotWorking', empty($path));
 
+// warn if the php module pcntl is missing for command line processing
+// not using function_exists('pcntl_signal') as this check would not run on the command line...
+$php_pcntl = exec('php -m | grep -i pcntl');
```
Comment (DeepDiver1975) [inline settings/admin.php]: no exec please

**Comment by DeepDiver1975** [comment]
I'd not consider this a relevant warning in here - print out on the occ command is enough from my understanding.

**Comment by oparoz** [comment]
>I'd not consider this a relevant warning in here - print out on the occ command is enough from my understanding.

@DeepDiver1975 - I agree, but wouldn't it be possible to have it at the global level instead of having each command which wants to be interruptible implement the check?

**Comment by karlitschek** [comment]
i agree with @deepdiver1975 we should move this to occ only

**Comment by DeepDiver1975** [comment]
> but wouldn't it be possible to have it at the global level instead of having each command which wants to be interruptible implement the check?

we can move this to the command base class which we use to implement all our commands.

**Comment by mmattel** [comment]
@DeepDiver1975 
to check if an optional php module is present on the command line, I need to use ``exec``, as function_exists('pcntl_signal') via the web environment may not respond with the right result.
If this is not wanted, we can close this PR as I do not see any chance to get a automated result (warning, info, whatever) we can show, based if php pcntl from the command line is avaliable or not.
I have already made a documentation PR where I added a statement about this optional php module. If this is enough - fine, but I thought it may be worth, based on the experience of @PVince81, to give the admin a notification about his setup.
Any other ideas are welcomed.

**Comment by DeepDiver1975** [comment]
> to check if an optional php module is present on the command line, I need to use exec, as function_exists('pcntl_signal') via the web environment may not respond with the right result.
> If this is not wanted, we can close this PR as I do not see any chance to get a automated result (warning, info, whatever) we can show, based if php pcntl from the command line is avaliable or not.

I fully understand that - but we really have to avoid exec.
As stated above it would be enough from my perspective to display a warning about this ins the start up of occ.

**Comment by mmattel** [comment]
ok, got that.
> we can move this to the command base class which we use to implement all our commands

My knowledge is currently not at the level to move the interruption logic to the command base class (beside some time constraints), so I please someone else for this task.

It would be great to have central reusable functions for command line interruption processing available, because we then just call them in case of needs. Not all occ functions may need process interrupt abilities, but in case of they could do more easily and it without reinventing the wheels...

For me it is ok to close this PR.

**Comment by DeepDiver1975** [comment]
here we go - https://github.com/owncloud/core/pull/22023

**Pull request closed without merging**